### PR TITLE
rework feat-cmake_tests CMake and replace scripts

### PR DESCRIPTION
**Pull request by byrnHDF**
Only the repack tests fail on local linux system.

Enable CMake testing:

- [ ] Convert Linux shell tests to Powershell to work on windows
- [x] Fix tests to build on windows
- [ ] Fix tests to pass on windows
- [ ] Enable github actions to run CMake tests on Linux
- [ ] Enable github actions to run CMake tests on Windows
- [ ] Verify the CMake tests cover the same as Autotool's tests

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -51,12 +103,30 @@ endif()
...
+if (NOT ZFP_FOUND)
+  message (FATAL_ERROR " ZFP is Required for h5z_zfp")
+endif ()
+
+#-----------------------------------------------------------------------------
+# Dashboard and Testing Settings
+#-----------------------------------------------------------------------------
+option (BUILD_TESTING "Build h5z-zfp Unit Testing" OFF)
+if (BUILD_TESTING)
+  set (DART_TESTING_TIMEOUT 1200 CACHE STRING
```
Comment (markcmiller86) [inline CMakeLists.txt:115]: What is `DART` in this variable name? Related to a CTest dashboard of some kind?

File: docs/cmake.rst [hunk 2]
```diff
@@ -23,7 +23,7 @@ Once you have installed both HDF5 and ZFP, you can compile H5Z-ZFP_ using a comm
 
 ::
 
-    export HDF_DIR=<path-to_hdf5>
+    export HDF5_DIR=<path-to_hdf5>
```
Comment (markcmiller86) [inline docs/cmake.rst:26]: dash vs. underscore?
```suggestion
    export HDF5_DIR=<path-to-hdf5>
```

File: test/CMakeLists.txt
Comment (markcmiller86) [inline test/CMakeLists.txt:1]: This is awesome 💪🏻 !

File: test/test_error.c [hunk 3]
```diff
@@ -5,19 +5,51 @@ Written by Mark C. Miller, miller86@llnl.gov
...
 #include <string.h>
-#include <strings.h>
+#if defined(_WIN32)
+#define _USE_MATH_DEFINES
+#include <math.h>
+#include <io.h>
+#define strncasecmp _strnicmp
+
+#include <ctype.h>
+char * strcasestr(s, find)
```
Comment (markcmiller86) [inline test/test_error.c:27]: Not sure if it applies here but In VisIt, I see `strcasestr` for Windows being handled as...

```
#ifdef _MSC_VER
    #include<shlwapi.h>
    StrStrIA(variableNames[0].c_str(), "coord"))
#else
    strcasestr(variableNames[0].c_str(), "coord"))
#endif
```

File: test/test_error.c [hunk 4]
```diff
@@ -251,9 +298,20 @@ int main(int argc, char **argv)
     if (0 > H5Fclose(fid)) ERROR(H5Fclose);
 
     /* Use raw file I/O to corrupt the dataset named corrupted_data */
+
```
Comment (markcmiller86) [inline test/test_error.c:301]: I guess we could simplify but just building the corrupted data once and committing it as a data file to be used in testing. Thoughts?

File: test/test_error.c [hunk 5]
```diff
@@ -251,9 +298,20 @@ int main(int argc, char **argv)
...
+#if defined(_WIN32)
+    FILE *fp;
+    errno_t err  = fopen_s( &fp, FNAME, "rb+");
+    if(err != 0) ERROR(fopen_s);
+    fseek(fp, (off_t) off + (off_t) siz / 3, SEEK_SET);
+    fwrite(corrupt, 1 , sizeof(corrupt), fp);
+    fclose(fp);
+#else
+    int fd;
```
Comment (markcmiller86) [inline test/test_error.c:310]: I don't think we really *need* to use `pwrite` here. We can just replace entirely with logic specified above which I believe should work also on Linux/macOS

File: test/test_read.c [hunk 6]
```diff
@@ -62,13 +62,50 @@ LLC,  and shall  not be  used for  advertising or  product endorsement
...
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
 
+
+#if defined(_WIN32)
+#include <ctype.h>
+#define strncasecmp _strnicmp
+// strcasestr is not available on Windows
+char * strcasestr(s, find)
```
Comment (markcmiller86) [inline test/test_read.c:79]: See not above about possible simpler alternative

File: test/test_write.c [hunk 7]
```diff
@@ -5,18 +5,67 @@ Written by Mark C. Miller, miller86@llnl.gov
...
+#define strncasecmp _strnicmp
+#define strcasecmp _stricmp
+#define read _read
+#define open _open
+#define close _close
+#define srandom(X) srand(X)
+#define random rand
+
+// strcasestr is not available on Windows
+#include <ctype.h>
```
Comment (markcmiller86) [inline test/test_write.c:35]: See note above.

I guess we could move all this code which is used similiarly in the various tests harness `.c` files to a common place. Not suggestion *you* should do that...only asking for opinions about doing that.

File: CMakeLists.txt [hunk 8]
```diff
@@ -20,6 +36,29 @@ set(H5Z_ZFP_VERSION "${H5Z_ZFP_VERSION_MAJOR}.${H5Z_ZFP_VERSION_MINOR}.${H5Z_ZFP
...
+    ${PROJECT_BINARY_DIR}/bin CACHE PATH "Single Directory for all static libraries."
+)
+set (CMAKE_Fortran_MODULE_DIRECTORY
+    ${PROJECT_BINARY_DIR}/mod CACHE PATH "Single Directory for all fortran modules."
+)
+get_property(_isMultiConfig GLOBAL PROPERTY GENERATOR_IS_MULTI_CONFIG)
+if(_isMultiConfig)
+  set (CMAKE_TEST_OUTPUT_DIRECTORY ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/${CMAKE_BUILD_TYPE})
+  set (CMAKE_PDB_OUTPUT_DIRECTORY
+      ${PROJECT_BINARY_DIR}/bin CACHE PATH "Single Directory for all pdb files."
```
[the remark this answers is not in the document]
Comment (byrnHDF) [reply CMakeLists.txt:56]: No, Microsoft uses pdb extension for debug symbols.

File: CMakeLists.txt [hunk 9]
```diff
@@ -38,11 +77,24 @@ set(CMAKE_INSTALL_CMAKEDIR lib/cmake/h5z_zfp)
...
-set(COMPONENTS C)
+set(COMPONENTS C shared)
 if (FORTRAN_INTERFACE)
   list(APPEND COMPONENTS Fortran)
 endif ()
-find_package(HDF5 REQUIRED COMPONENTS ${COMPONENTS})
+# By default static will be returned unless only shared is available
+find_package(HDF5 REQUIRED COMPONENTS ${COMPONENTS} CONFIG)
+if (NOT HDF5_FOUND)
+  message (FATAL_ERROR " HDF5 is Required for h5z_zfp")
```
[the remark this answers is not in the document]
Comment (byrnHDF) [reply CMakeLists.txt:87]: fixed.

File: CMakeLists.txt [hunk 10]
```diff
@@ -51,12 +103,30 @@ endif()
 
 # Find ZFP, relies on ZFP_DIR or ZFP_ROOT being set in environment.
 find_package(ZFP REQUIRED CONFIG)
+if (NOT ZFP_FOUND)
+  message (FATAL_ERROR " ZFP is Required for h5z_zfp")
```
[the remark this answers is not in the document]
Comment (byrnHDF) [reply CMakeLists.txt:107]: fixed.

File: CMakeLists.txt (hunk 1 above)
Comment (byrnHDF) [reply CMakeLists.txt:115]: DART_TESTING_TIMEOUT is the CMake variable name that must be set to timebox tests. (yes, it refers to an old CTest dashboard)

File: docs/cmake.rst (hunk 2 above)
Comment (byrnHDF) [reply docs/cmake.rst:26]: dashes are consistent

**Comment by byrnHDF** [comment]
Will leave test code comments to Scot.

**Comment by byrnHDF** [comment]
There is a macro that need to be added that will allow the various sources of hdf5 binaries to be used. This PR only checks for CMake built HDF5. I have a macro in use by other projects that will search for variations.

**Comment by markcmiller86** [comment]
@byrnHDF how does this PR relate to #110

**Comment by markcmiller86** [comment]
Looks like GCC build is having trouble finding HDF5 package.

**Comment by byrnHDF** [comment]
> Looks like GCC build is having trouble finding HDF5 package.

Right - that is the need for the macro. To find hdf5 regardless of how it was built (autotools or two different CMake ways)

**Comment by markcmiller86** [comment]
> Right - that is the need for the macro. To find hdf5 regardless of how it was built (autotools or two different CMake ways)

Ok, so does that mean that we're still waiting on those finishing touches here in this PR?

**Comment by byrnHDF** [comment]
Correct. This PR gives a preview of the changes compared to Scot feat-cmake_tests PR. (For review)
Plus it allows easier access to the code from other platforms (Like windows) for testing.

**Comment by byrnHDF** [comment]
Just added a macro that should now find the variety of ways that hdf5 is presented.
Windows still needs debugging.

**Comment by markcmiller86** [comment]
Note...I am noticing I or someone else who has necessary privs here winds up having to hit the `Approve and run` button for CI checks on a PR from a fork. I think that is because we're triggering with `pull_request` instead of `pull_request_target`... here, https://github.com/LLNL/H5Z-ZFP/blob/1660d94a4f73a350d308e262f5ec022cc896a9c4/.github/workflows/main.yml#L8

ChatGPT explains the difference...

> What is the difference between the pull_request and pull_request_target triggers in GitHub actions?

In GitHub Actions, pull_request and pull_request_target are both event triggers that are used to run workflows in response to pull request events. However, they have some differences in how they work.

The pull_request trigger is activated whenever a pull request is opened, reopened, or synchronized (when a new commit is added to the branch that the pull request is based on). This trigger runs the workflow on the head repository (i.e., the repository where the changes were made) and provides information about the pull request and its associated branch.

The pull_request_target trigger, on the other hand, is activated when a pull request is opened or synchronized, but it runs the workflow on the base repository (i.e., the repository where the pull request is being merged into). This trigger provides information about the pull request, its associated branch, and the repository that it is being merged into.

The key difference between the two triggers is that pull_request_target is more secure as it only runs when a pull request is opened or synchronized from a forked repository into the base repository. It ensures that the workflow is only triggered on the base repository, even if the pull request comes from an untrusted source.

In summary, if you are running a workflow that needs access to secrets or other sensitive information, it is recommended to use pull_request_target instead of pull_request to ensure security.

**Comment by brtnfld** [comment]
Github has a requirement that first-time contributors will require approval to run workflows. It's in the actions settings for the repo.

File: test/test_write.c (hunk 7 above)
Comment (brtnfld) [reply test/test_write.c:35]: I think it would be a good idea to combine the common routines to one place.

File: test/test_error.c (hunk 5 above)
Comment (brtnfld) [reply test/test_error.c:310]: Sounds good, I'll update it.

File: test/test_error.c (hunk 4 above)
Comment (brtnfld) [reply test/test_error.c:301]: It is nice to have the program around that generated the file, I don't necessarily have a problem generating in on the fly.

File: test/test_write.c (hunk 7 above)
Comment (brtnfld) [reply test/test_write.c:35]: See PR #112

File: test/test_error.c (hunk 5 above)
Comment (brtnfld) [reply test/test_error.c:310]: See PR #112

File: test/test_error.c (hunk 3 above)
Comment (brtnfld) [reply test/test_error.c:27]: See PR #112
